util/sst_file_manager_impl.cc

```diff
@@ -166,6 +227,135 @@ uint64_t SstFileManagerImpl::GetTotalTrashSize() {
   return delete_scheduler_.GetTotalTrashSize();
 }
 
+void SstFileManagerImpl::ReserveDiskBuffer(uint64_t size,
+                                           const std::string& path) {
+  MutexLock l(&mu_);
+
+  reserved_disk_buffer_ += size;
+  if (path_.empty()) {
+    path_ = path;
+  }
+}
+
+void SstFileManagerImpl::ClearError() {
+  while (true) {
+    MutexLock l(&mu_);
+
+    while (error_handler_list_.empty() && !closing_) {
+      cv_.Wait();
+    }
+
+    if (closing_) {
+      return;
+    }
+
+    uint64_t free_space;
+    Status s = env_->GetFreeSpace(path_, &free_space);
+    if (s.ok()) {
+      if (bg_err_.severity() == Status::Severity::kHardError) {
+        // Give priority to hard errors
```

Comment:
Can you comment more about what it means?

Reply:
Its referring to the case where some DB instances experienced soft errors (i.e compaction failure) and some experienced hard errors (i.e flush/WAL write failures), we would estimate required free space as if all instances experienced hard errors. For hard errors, the free space threshold is ```reserved_disk_buffer_```, which is basically the sum of write_buffer_size of all instances. For soft errors, the threshold is ```free_space_trigger_```, which is the value of ```cur_compactions_reserved_size_``` at the time of error.